File: iOS/Services/DiagnosticsReporter.swift
@preconcurrency import FirebaseAnalytics
@preconcurrency import FirebaseCrashlytics
import Foundation

public enum DiagnosticsPrivacyPolicy {
    private static let allowedEvents: Set<String> = [
        "plan_access_view",
        "billing_products_loaded",
        "billing_purchase_started",
        "billing_purchase_result",
        "billing_restore_started",
        "billing_restore_result",
        "activation_milestone"
    ]
    private static let commercialKeys: Set<String> = ["plan", "billing_period", "source", "product_id", "result"]
    private static let allowedKeysByEvent: [String: Set<String>] = Dictionary(
        uniqueKeysWithValues: allowedEvents.map { event in
            (event, event == "activation_milestone" ? Set(["milestone", "state"]) : commercialKeys)
        }
    )

    public static func sanitize(
        event: String,
        fields: [String: String]
    ) -> (String, [String: String])? {
        guard let allowedKeys = allowedKeysByEvent[event] else { return nil }
        let safeFields = fields.reduce(into: [String: String]()) { result, item in
            guard allowedKeys.contains(item.key) else { return }
            result[item.key] = sanitizeValue(item.value)
        }
        return (event, safeFields)
    }

    private static func sanitizeValue(_ value: String) -> String {
        let lowered = value.lowercased()
        let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyz0123456789_-")
        let normalized = lowered.unicodeScalars.map { allowed.contains($0) ? Character(String($0)) : "_" }
        return String(normalized.prefix(40))
    }
}

public enum CommercialTelemetryFields {
    public static func product(_ productId: String, source: String) -> [String: String] {
        var fields = ["product_id": productId, "source": source]
        guard let product = CommercialProductCatalog.products.first(where: { $0.productId == productId }) else {
            return fields
        }
        fields["plan"] = product.plan.rawValue
        fields["billing_period"] = billingPeriod(product.billingPeriod)
        return fields
    }

    public static func result(_ result: String, source: String, productId: String? = nil) -> [String: String] {
        var fields = productId.map { product($0, source: source) } ?? ["source": source]
        fields["result"] = result
        return fields
    }

    private static func billingPeriod(_ period: BillingPeriod) -> String {
        switch period {
        case .monthly: return "monthly"
        case .annual: return "annual"
        }
    }
}

public enum DiagnosticsReporter {
    private static let preferenceKey = "shareAnonymousDiagnostics"

    public static var isEnabled: Bool {
        UserDefaults.standard.bool(forKey: preferenceKey)
    }

    public static func applyStoredConsent() {
        setCollection(isEnabled)
    }

    public static func setConsent(_ enabled: Bool) {
        UserDefaults.standard.set(enabled, forKey: preferenceKey)
        setCollection(enabled)
    }

    public static func event(_ name: String, fields: [String: String] = [:]) {
        guard isEnabled else { return }
        guard let sanitized = DiagnosticsPrivacyPolicy.sanitize(event: name, fields: fields) else { return }
        let parameters = sanitized.1.reduce(into: [String: Any]()) { result, item in
            result[item.key] = item.value
        }
        Analytics.logEvent(sanitized.0, parameters: parameters)
    }

    private static func setCollection(_ enabled: Bool) {
        Analytics.setAnalyticsCollectionEnabled(enabled)
        let crashlytics = Crashlytics.crashlytics()
        crashlytics.setCrashlyticsCollectionEnabled(enabled)
        if !enabled { crashlytics.deleteUnsentReports() }
    }
}

File: Android/app/src/main/java/com/example/supplementtracker/service/DiagnosticsReporter.kt
package com.example.supplementtracker.service

import android.content.Context
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics

object DiagnosticsPrivacyPolicy {
    private val allowedEvents = setOf(
        "plan_access_view",
        "billing_purchase_started",
        "billing_restore_started"
    )
    private val allowedKeys = setOf("plan", "billing_period", "source")

    fun sanitize(event: String, fields: Map<String, String>): Pair<String, Map<String, String>>? {
        if (event !in allowedEvents) return null
        val safeFields = fields.filterKeys { it in allowedKeys }
            .mapValues { (_, value) -> sanitizeValue(value) }
        return event to safeFields
    }

    private fun sanitizeValue(value: String): String {
        return value.lowercase().replace(Regex("[^a-z0-9_-]"), "_").take(40)
    }
}

object DiagnosticsReporter {
    private const val preferenceKey = "shareAnonymousDiagnostics"

    fun applyStoredConsent(context: Context) {
        setCollection(context, OakPrefs.get(context).getBoolean(preferenceKey, false))
    }

    fun setConsent(context: Context, enabled: Boolean) {
        OakPrefs.get(context).edit().putBoolean(preferenceKey, enabled).apply()
        setCollection(context, enabled)
    }

    fun isEnabled(context: Context): Boolean {
        return OakPrefs.get(context).getBoolean(preferenceKey, false)
    }

    fun event(context: Context, name: String, fields: Map<String, String> = emptyMap()) {
        if (!isEnabled(context)) return
        val sanitized = DiagnosticsPrivacyPolicy.sanitize(name, fields) ?: return
        FirebaseAnalytics.getInstance(context).logEvent(sanitized.first, sanitized.second.toBundle())
    }

    private fun setCollection(context: Context, enabled: Boolean) {
        FirebaseAnalytics.getInstance(context).setAnalyticsCollectionEnabled(enabled)
        val crashlytics = FirebaseCrashlytics.getInstance()
        crashlytics.setCrashlyticsCollectionEnabled(enabled)
        if (!enabled) crashlytics.deleteUnsentReports()
    }

    private fun Map<String, String>.toBundle(): Bundle = Bundle().also { bundle ->
        forEach { (key, value) -> bundle.putString(key, value) }
    }
}
